### PR TITLE
Interleave commandline require/set/unset commands

### PR DESCRIPTION
**Pull request by LasseBlaauwbroek**
This is an initial implementation of the interleaving of command-line require/set/unset commands as discussed in #11851. The purpose of this is to allow people to set options that have been defined in a plugin loaded by a previous `-require-from` command. Some notes:

- In #11851 it was said by @ejgallego that `-load-vernac-object` behaves differently from the `-require-*` family and should be deprecated. I cannot find any evidence of this in the code so I have not done this. Please let me know if I missed where this difference appears.
- It seems to me that in addition to `-require-*` and `-set`/`-unset`, the command `-load-vernac-source` should also be interleaved for consistency. However, when trying to implement this, I ran into difficulties. The `-require-*` commands are processed during STM initialization, while `-load-vernac-source` happens after initialization (in `ccompile.ml`). So either one of these should be moved, but this does not seem trivial. Guidance on this would be appreciated.
- Finally, it occurs to me that there is another design option: One could deprecate the `-set`/`-unset` options, and allow `-load-vernac-source` option to load files without a `.v` extension. Then setting and unsetting options could be accomplished on the command line using process substitution like this:
```bash
coqtop -load-vernac-source <(<<< "Set My Option")
```
Currently, this approach is not possible because Coq won't load the temporary file without a `.v` extension that is created.

**Kind:** feature.

- [ ] Added / updated test-suite

- [ ] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [ ] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Review by ejgallego — COMMENTED** [review]
Thanks @LasseBlaauwbroek , after a quick review this looks pretty good I'd say.

I'll try to leave some more detailed comments ASAP while other can have a look if they wish; but in general I think I have only minor comments so at most one revision may be necessary.

**Comment by Zimmi48** [comment]
>In #11851 it was said by @ejgallego that `-load-vernac-object` behaves differently from the `-require-*` family and should be deprecated. I cannot find any evidence of this in the code so I have not done this. Please let me know if I missed where this difference appears.

In fact, he probably meant `-load-vernac-source` which, as you've noticed, is handled separately and will be harder to change.

>Finally, it occurs to me that there is another design option: One could deprecate the `-set`/`-unset` options, and allow `-load-vernac-source` option to load files without a `.v` extension. Then setting and unsetting options could be accomplished on the command line using process substitution like this:
> 
> 
> ```shell
> coqtop -load-vernac-source <(<<< "Set My Option")
> ```
> 
> Currently, this approach is not possible because Coq won't load the temporary file without a `.v` extension that is created.

This is interesting but:
1. This would be much dependent on shell syntax, which is not great for portability.
2. It would also be more complex to use, whereas `-set` / `-unset` options are pretty simple.
3. With this solution, there would be no way to set options before some plugin is required.

That being said, I'd say it would make sense dropping the requirement for a `.v` extension with `-load-vernac-source` when a filename is provided. (This is quite independent of this PR, so if you want to propose this change, please open a separate PR.)

**Review by Zimmi48 — COMMENTED** [review]
This looks pretty good to me too :smile: 

I feel like you could leave some comments in the code explaining why it is not done for `-load-vernac-source`.

Regarding the documentation, you will have to tweak it again and mention that `-load-vernac-source` are always processed *after* any `-set` and `-require-*` command.

Regarding the changelog, instead of creating a new entry, please edit your previous entry by adding the link to this PR and explaining only the new behavior, not the intermediate state.

**Comment by Zimmi48** [comment]
If you could, it would be great to add some tests. Tests in the test-suite can take a header like:
```coq
(* coq-prog-args: ("-impredicative-set") *)
```

to pass some arguments to `coqc`. (Unfortunately, this is not yet documented in the test-suite's [README](https://github.com/coq/coq/blob/master/test-suite/README.md) :disappointed: )

You could interleave some `-require-import` and some `-set` using some Coq options that are only defined in plugins that must be loaded. Unfortunately, you can only test success with this infrastructure, not failure.

**Comment by SkySkimmer** [comment]
>Unfortunately, you can only test success with this infrastructure, not failure.

You can also use a misc/ test which is arbitrary shell

File: stm/stm.mli [hunk 1]
```diff
@@ -52,6 +52,19 @@ type stm_doc_type =
...
+
+type injection_command =
+  | OptionInjection of (Goptions.option_name * option_command)
+  (** Require libraries before the initial state is
+     ready. Parameters follow [Library], that is to say,
+     [lib,prefix,import_export] means require library [lib] from
+     optional [prefix] and [import_export] if [Some false/Some true]
+     is used.  *)
+  | RequireInjection of (string * string option * bool option)
+  (* None = No Import; Some false = Import; Some true = Export *)
```
Comment (ejgallego) [inline stm/stm.mli:65]: If you are using `(** `  above to have a comment docstring do so here too.

File: stm/stm.mli [hunk 2]
```diff
@@ -66,12 +79,9 @@ type stm_init_options =
...
      options / _CoqProject *)
 
-  ; require_libs : (string * string option * bool option) list
-  (** Require [require_libs] before the initial state is
-     ready. Parameters follow [Library], that is to say,
-     [lib,prefix,import_export] means require library [lib] from
-     optional [prefix] and [import_export] if [Some false/Some true]
-     is used.  *)
+  ; injections : injection_command list
+  (* Injects Load, Require and Set/Unset commands before the initial
```
Comment (ejgallego) [inline stm/stm.mli:83]: Same problem with doctring, in this case you remove a valid one.

File: stm/stm.mli [hunk 3]
```diff
@@ -66,12 +79,9 @@ type stm_init_options =
   (** [vo] load paths for the document. Usually extracted from -R
      options / _CoqProject *)
 
-  ; require_libs : (string * string option * bool option) list
-  (** Require [require_libs] before the initial state is
-     ready. Parameters follow [Library], that is to say,
-     [lib,prefix,import_export] means require library [lib] from
-     optional [prefix] and [import_export] if [Some false/Some true]
-     is used.  *)
+  ; injections : injection_command list
```
Comment (ejgallego) [inline stm/stm.mli:82]: Not sure about the name `injections` , I cannot come with a much better one tho; maybe `startup_params` ?

**Review by ejgallego — CHANGES_REQUESTED** [review]
Other than the comment on naming this is good modulo documentation:
- a changelog is needed, see `dev/tools/make-changelog.sh`, or even better update the previous one as suggested by @Zimmi48 
- a modification of the manual seems indeed needed too
- the in-code docstrings are broken.
- a test would be convenient
Once that is fixed we can merge.

File: stm/stm.mli (hunk 3 above)
Comment (Zimmi48) [reply stm/stm.mli:82]: I kind of like that "injection" name.

**Comment by Zimmi48** [comment]
> a changelog is needed, see `dev/tools/make-changelog.sh`

In this case though, I'd suggest modifying the changelog entry of the previous PR instead of creating a new one.

**Comment by Zimmi48** [comment]
@LasseBlaauwbroek I'd like to ensure this gets merged in time for 8.12 (see the recent announcement regarding the release schedule: https://coq.discourse.group/t/8-12-release-schedule/809). I can take care of finishing your PR, but I need you to let me know if that's something you'd like of if you'll get some time to finish it yourself.

**Comment by ejgallego** [comment]
Thanks @Zimmi48 , I was planning myself to push to the PR, but that would require changing assignee so I'd be great if you can do it and I'll merge.

**Comment by LasseBlaauwbroek** [comment]
Yes, sorry I've been busy with other things and this slipped through. I'll update the changelog and other minor things today. I'm not sure if I'll manage to do the tests though.

**Comment by Zimmi48** [comment]
> Yes, sorry I've been busy with other things and this slipped through. I'll update the changelog and other minor things today. I'm not sure if I'll manage to do the tests though.

OK thanks! Let us know and we'll take care of what you didn't manage to do by yourself.

**Comment by Zimmi48** [comment]
I'm taking care of this PR now.

**Comment by Zimmi48** [comment]
Should be ready now (if CI agrees).

**Comment by ejgallego** [comment]
Still a problem with update-compat.sh

**Comment by Zimmi48** [comment]
Oh right! Sorry, I didn't check this part of the test-suite.

**Comment by Zimmi48** [comment]
The cause of the issue was an additional commit that I had included in the PR but that was slightly orthogonal, so I've now removed it.

**Review by ejgallego — APPROVED** [review]
LGTM, thanks to all!

**Pull request merged**